### DISCUSSION TITLE
Improving the Typst base template to match equivalent PDF options

### DISCUSSION THREAD

**Discussion by christopherkenny** [quarto-dev/quarto-cli#10223] [Feature Requests]
### Description

I've had to create several new template classes for my work in Quarto recently. I made the switch to using Typst for all of them, but still using Quarto as the frontend.

As a heavy Quarto + LaTeX user prior, it's clear that there are many fewer features available for `-typst` formats compared to `-pdf` formats. It looks like most of this is driven by (1) the relative youth of Typst (~5) compared to LaTeX (~40) and (2) the simplicity of the upstream Pandoc template.

Typst itself is a very intuitive alternative to LaTeX and (at least to me) is substantially easier to learn and use than LaTeX ever was/is. As such, I think it would be great if some features of the Typst templates could be caught up to `-pdf` outputs.
To this end, I've listed out many features where the implementation would be only a few lines and thus hopefully easy to maintain. Of course, in aggregate, this would add quite a few features at once.

I'm starting here with (1) default suggestions, (2) example syntax for how it would be implemented based on the YAML variables, and (3) details explaining the rationale behind any choices. If there are changes that would be helpful, I'm happy to make a (series of) pull requests to introduce these features to Quarto's templates.

One important thing to note: Some of these examples may need to be prefixed by `#` to start code while others may need to have `#` removed. Each of those is context-dependent on where the "best" place to insert such a fix would be. The rest of the content would stay the same.

Finally, I realize this is a long list of possibilities.
If you're not open to a pull request on some (or any) of these points, then this discussion just documents how to get the "equivalent" behaviors in Typst formats as in the other pdf formats.

The rest of this proceeds as follows:

- Introduces a general comment on passing variables Quarto -> Pandoc -> Typst.
- Compares the [PDF options](https://quarto.org/docs/reference/formats/pdf.html) and [Typst options](https://quarto.org/docs/reference/formats/typst.html). Sections are organized following headings on the PDF options page.
- Discusses the layout options that would require more work.

## Quarto -> Pandoc -> Typst

[Pandoc aggressively escapes characters](https://github.com/jgm/pandoc/issues/9823) that can cause issues for Pandoc's processing, but would be valid in Typst. This can make it hard to pass things like colors, emails, or urls from Quarto to Typst. Specifically, if you pass a variable from Pandoc (say $my-variable$) to Typst, then $my-variable$ gets escaped. So, if you pass `email: chris@example.com` and try to make "$email$", it becomes `chris\@example.com`. For things like emails, a simple [Typst-evaluated regex](https://typst.app/docs/reference/foundations/regex/) might suffice, but it falls apart if you don't know all possible escapes.

Instead, if you make `[$email$]`, it passes it as "context". This avoids nasty escape characters. Yet, context is messier to work with. The easiest way to extract text from context is a function like:

```typst
#let to-string(content) = {
  if content.has("text") {
    content.text
  } else if content.has("children") {
    content.children.map(to-string).join("")
  } else if content.has("body") {
    to-string(content.body)
  } else if content == [ ] {
    " "
  }
}
```

This short function comes from an [old issue](https://github.com/typst/typst/issues/2196) but handles this problem very well.

Adding a function like this would also fix printing errors in templates, like [the IEEE one](https://github.com/quarto-ext/typst-templates/tree/main/ieee). If you render that template, you can see how the Pandoc escaping makes the email links print incorrectly and have invalid links.

#### `subtitle`

Default: `none`

```typst
if title != none {
    if subtitle != none {
        align(center)[#block(inset: 2em)[
            #text(weight: "bold", size: 1.5em)[#title]
            #parbreak()
            #text(weight: "bold", size: 1.25em)[#subtitle]
        ]]
        } else {
            align(center)[#block(inset: 2em)[
            #text(weight: "bold", size: 1.5em)[#title]
        ]]
    }
}
```

This example solution is from a [prior discussion](https://github.com/quarto-dev/quarto-cli/discussions/9638) and was made by `@mcanouil`.

#### `thanks`

Default: `none`

```typst
if title != none {
    align(center)[#block(inset: 2em)[
        #text(weight: "bold", size: 1.5em)[
            #title #if thanks != none {
                footnote(thanks, numbering: "*")
                counter(footnote).update(n => n - 1)
            }
        ]
    ]]
}
```
Details: This may actually be one of the most complicated suggestions, as it interacts with the title and a counter. However, this is just adding a footnote and using the "*" numbering. We then reset the counter to 0, which could be as simple as `counter(footnote).update(0)`. In case something gets stuck before the title, we can use `n => n - 1`.

#### `linestretch`

Can be specified as any non-negative number.

Default: `1` (single spacing)

```typst
set par(leading: linestretch * 0.65em)
```

Details: 0.65em is the measurement for one line.

#### `mathfont`

Default: New Computer Modern Math

```typst
#show math.equation: set text(font: mathfont)
```

Details: New Computer Modern Math is shipped with Typst.

##### `codefont`

Default: DejaVu Sans Mono (shipped with Typst)

```typst
#show raw: set text(font: codefont)
```

Details: DejaVu Sans Mono is shipped with Typst.

#### `linkcolor`

Default: "#800000"

```typst
#show link: set text(fill: rgb(to-string(linkcolor)))
```
Details: Matches Quarto's PDF template default.
Related discussion: https://github.com/quarto-dev/quarto-cli/discussions/9175

#### `filecolor`, `urlcolor`

Default: "#800000"

```typst
#show link: this => {
    if type(this.dest) != label {
        text(this, fill: rgb(to-string(filecolor)))
    }
}
```

Details: Try to match Quarto's PDF template default, but not possible in a simple way.

Note: These can't be separated without some regex, as far as I know.

#### `citecolor`

Default: "#0000FF"

```typst
#show ref: this => {
    text(this, fill: rgb(to-string(citecolor)))
}
```

Details: Try to match Quarto's PDF template default, but not possible in a simple way.
This catches both crossreferences and bibliographic references.

Note: These can't be separated, as far as I know.

#### `toccolor`

Default: "#0000FF"

```typst
#show link: this => {
    if type(this.dest) == label {
        text(this, fill: rgb(to-string(filecolor)))
    }
}
```
Details: Not 100% sure the default, but this looks right to me from the LaTeX.

#### `keywords`

Default: [don't set]

```typst
#set document(keywords: keywords)
```
Details: Defaults to an empty array in Typst.

#### `title-meta`

Default: `title` (or [don't set])

```typst
#set document(title: title-meta)
```

Details: Defaults to `none` in Typst.

#### `author-meta`

Default: [don't set]

```typst
#set document(author: author-meta)
```
Details: Defaults to an empty array in Typst.

#### `date-meta`

Default: `auto`

```typst
#set document(date: date-meta)
```
Details: Set by Typst to current date by default

### layout

These have very simple adjustments in theory.
Each fix is a line or two, but they are currently hardcoded when converted from Quarto to Typst.
For example, see this [line](https://github.com/quarto-dev/quarto-cli/blob/b7150a0d767cee7d0aac1796bfe57847b3c65160/src/resources/filters/layout/pandoc3_figure.lua#L163).

#### `cap-location`

Problem: This is hardcoded when creating the figures and tables.
Otherwise, changing the default *should be* as simple as combining both the table and figure suggestions.

```typst
#show figure.where(kind: "quarto-float-fig"): set figure.caption(position: cap-location)
#show figure.where(kind: "quarto-float-tbl"): set figure.caption(position: cap-location)
```

#### `fig-cap-location`

Problem: This is hardcoded when creating the figures.
Otherwise, changing the default *should be* a single line.

Default: "bottom"

```typst
#show figure.where(kind: "quarto-float-fig"): set figure.caption(position: fig-cap-location)
```

Note: Also listed in Figures in the PDF Options document.

#### `tbl-cap-location`

Problem: This is also hardcoded when creating the figures.
Otherwise, changing the default *should be* a single line.

Default: "top"

```typst
#show figure.where(kind: "quarto-float-tbl"): set figure.caption(position: tbl-cap-location)
```

Note: Also listed in Tables in the PDF Options document.

**Comment by mcanouil** [comment]
Thanks for the suggestions.

Note that a lot of the YAML options come directly from Pandoc and the template there.
For instance, Quarto updates Pandoc thus the templates at each new Quarto releases.

Also, I believe most of your suggestions should be applied on Pandoc side as I did for subtitle.

cc. @gordonwoodhull 

FYI:
- https://github.com/quarto-dev/quarto-cli/issues/10212

**Reply by christopherkenny** [reply]
After some hiccups and making too large a PR to close quickly, many of these features are now added on the Pandoc side: https://github.com/jgm/pandoc/pull/9970 and https://github.com/jgm/pandoc/issues/11076 cover them.

**Reply by the-solipsist** [reply]
Thanks for this, @christopherkenny !

**Comment by gordonwoodhull** [comment]
This is a really helpful catalog of areas where Typst output could be brought up to the level of Latex. Thanks!

I agree with @mcanouil that most of these changes should be made on the Pandoc side. For instance, we will pick up his `subtitle` fix when we upgrade to Pandoc 3.2.1+ in Quarto 1.6.

I would suggest contributing the Pandoc features directly to Pandoc!

I am not sure, but I think the "more difficult layout problems" may require changes on the Quarto side. If you would be willing to do a little more background research to see where these changes need to be made, that would be super helpful.

Please keep in mind that while we have a long-term plan for how to review and accept changes that improve feature parity between formats, that is at least some months away. (Among other things, we need guidelines and processes on how to add tests, docs, and feature-format matrix entries.)

**Comment by christopherkenny** [comment]
Thank you both for the feedback. I can start an issue at Pandoc (edit: see https://github.com/jgm/pandoc/issues/9956) that goes through the Pandoc pieces of this. I was not entirely clear how much templates differ between Pandoc and Quarto.

My sense is that the layout problems would require Quarto changes, but I will look into how much is required there. Notably, since Quarto uses its own figure type, then any layout directives need to be handled here, separately from Pandoc. (This is because you have to target them with `figure.where(kind: "quarto-float-***")`, as Quarto does not insert standard "table" or "image" kinds.)

**Reply by mcanouil** [reply]
> I was not entirely clear how much templates differ between Pandoc and Quarto.

The idea is to have the smallest amount of differences to ease the process of Pandoc update.

> Notably, since Quarto uses its own figure type, then any layout directives need to be handled here, separately from Pandoc. (This is because you have to target them with figure.where(kind: "quarto-float-***"), as Quarto does not insert standard "table" or "image" kinds.)

To be accurate, it's not figure. It's "whatever" including figures and tables.
See <https://quarto.org/docs/authoring/cross-references-custom.html>.

**Reply by christopherkenny** [reply]
Thank you, that's helpful to know.

**Comment by cscheid** [comment]
Hi @christopherkenny, thanks for the long post.

Let me add a bit of context to some of our design goals and decisions, so that you understand where we might land on eventual PRs you might make. Then I'll make some specific comments.

- We prefer "features" that work uniformly across formats
- We prefer our templates to feel familiar to users coming from "non-Quarto workflows" of that format
- We prefer our templates to feel familiar to users coming from "pure Pandoc workflows" for that format

These three goals mean that we want Quarto to work, in general, as close to a "strict Pandoc superset".

## Specific comments

> `cap-location` Problem: This is hardcoded when creating the figures and tables.

I'm not sure I understand: `cap-location` can very much be controlled by users authoring the document via YAML configuration. In case you mean something like "the control of a Typst feature should be possible in Typst", let me respond to that.

Quarto is very much designed in a way that doesn't prioritize the cleanliness of the intermediate output files. It's not that we're unwilling to make changes that make the output files cleaner; it's that we don't consider this to be particularly important. We instead choose to focus on uniformity of experience for Markdown and ipynb inputs, and will almost always prioritize that over cleanliness of output files.

Quarto _does_ intend the intermediate outputs to be consumable by downstream tooling without requiring Quarto to be installed. The typical use case here is that we want the intermediate latex files to exist and be available for submission to a journal that requires LaTeX sources. If (hopefuly when!) journals start accepting raw Typst as their format submission, we intend Quarto's Typst output to be directly usable as input to such submissions. But Quarto does not (and will not) ever emit Typst code that is as clean as you could write by hand.

## Next steps

I realize that this might mean that some of your proposed changes do not match our design goals.

I think the best way to incorporate your suggestions into Quarto is for you to start with the very smallest PR you can envision, so that we can together experiment with how that feature interacts with Quarto's other formats and our design goals.

**Reply by christopherkenny** [reply]
Thank you for the additional context. I've started [an issue upstream about the pandoc template](https://github.com/jgm/pandoc/issues/9956), which based on your comment and the others is the best course of action for any non-Quarto-specific edits. Unless you suggest otherwise, I think it might be wise to wait for any thoughts/considerations there before starting on any next steps here, given the feedback.

As for the `cap-location` point, I think I goofed. Sorry about that. I'm not sure exactly what I did, but I convinced myself that `cap-location` and the `tab/fig-cap-location` arguments did nothing. I'm glad those arguments work natively.

It may be worth clarifying the "hardcoded" comment. My best understanding from tracing though the code is:
- If no specific yaml position is given for the caption when a figure is created (in the .qmd file), then this gets processed to the default [around here](https://github.com/quarto-dev/quarto-cli/blob/b7150a0d767cee7d0aac1796bfe57847b3c65160/src/resources/filters/customnodes/floatreftarget.lua#L130C1-L135C59)
  - The default will always take on a value because the last option is to take from `crossref`, [as defined starting here](https://github.com/quarto-dev/quarto-cli/blob/b7150a0d767cee7d0aac1796bfe57847b3c65160/src/resources/filters/mainstateinit.lua#L21).
- Whatever value is accessed at this point is the passed to `position:` within `figure.caption()` either [here for a figure with subfigures](https://github.com/quarto-dev/quarto-cli/blob/b7150a0d767cee7d0aac1796bfe57847b3c65160/src/resources/filters/layout/typst.lua#L138) or [here otherwise](https://github.com/quarto-dev/quarto-cli/blob/b7150a0d767cee7d0aac1796bfe57847b3c65160/src/resources/filters/layout/typst.lua#L147C26-L147C42)

In effect, this means that any creation of a caption will *always* be accompanied by either `position: top` or `position: bottom`.
Typst should *always* respect that value then, as far as I know, since that's directly setting an option above and beyond a set or show rule.

This is likely where the Typst-Quarto tradeoff is:
- Using Typst, I would want any `show-set` rules to apply to all figures that I have not explicitly said to do something else with.
- Using Quarto, I would want it to pick sensible defaults, which may require that it explicitly says to do something.

**Reply by cscheid** [reply]
> In effect, this means that any creation of a caption will _always_ be accompanied by either `position: top` or `position: bottom`. Typst should _always_ respect that value then, as far as I know, since that's directly setting an option above and beyond a set or show rule.
> 
> This is likely where the Typst-Quarto tradeoff is:
> 
> * Using Typst, I would want any `show-set` rules to apply to all figures that I have not explicitly said to do something else with.
> * Using Quarto, I would want it to pick sensible defaults, which may require that it explicitly says to do something.

That's correct, and this is an instance of what I meant by Quarto _not_ adhering to "the control of a Typst feature should be possible in Typst". If a feature can be controlled in the Markdown document (either through classes, attributes, or YAML metadata), then we _don't_ consider it a bug that the resulting behavior is "hardcoded" in Typst.

To quote my original response again, "Quarto is very much designed in a way that doesn't prioritize the cleanliness of the intermediate output files.". I will say something even stronger: Quarto considers the .typ output to not be suitable for humans to read or edit. That is by design, and allows us sufficiently many favorable tradeoffs that we are not going to budge on that.

**Reply by christopherkenny** [reply]
Thank you for the additional information on that. 

For what it's worth, I came across this behavior while working on a format extension which makes a PDF using Typst, not while trying to use the intermediate output for anything (but debugging). I found the behavior surprising, but this context is very helpful. Thanks.
